src/runtime/TemporalObject.cpp

```diff
@@ -68,228 +68,284 @@ Value TemporalObject::toISOTime(ExecutionState& state, DateObject& d)
     return Value(new ASCIIString(result.data(), result.length()));
 }
 
-std::map<std::string, std::string> TemporalObject::parseValidIso8601String(ExecutionState& state, const Value& str)
+
+std::string TemporalObject::getNNumberFromString(std::string& isoString, const int n)
 {
-    ASSERT(str.isString());
-    size_t pos = str.asString()->toNonGCUTF8StringData().find('[');
-    std::string isoDate = str.asString()->toNonGCUTF8StringData().substr(0, pos);
-    std::string extra;
+    std::string retVal;
+    if (!TemporalObject::is_number(isoString.substr(0, n))) {
+        throw std::exception();
+    }
+    retVal = isoString.substr(0, n);
+    isoString = isoString.substr(n);
+    return retVal;
+}
 
-    if (pos != std::string::npos) {
-        extra = str.asString()->toNonGCUTF8StringData().substr(pos);
+std::map<std::string, std::string> TemporalObject::getSeconds(std::string& isoString)
+{
+    int counter = 1;
+    while (TemporalObject::is_number(isoString.substr(counter, 1))) {
+        counter++;
+    }
+    if (counter == 1 || counter > 9) {
+        throw std::exception();
     }
 
-    pos = isoDate.find('T');
+    std::string tmp = isoString.substr(1, counter - 1);
+
+    isoString = isoString.substr(counter);
 
-    if (pos != std::string::npos) {
-        std::replace(isoDate.begin(), isoDate.end(), 'T', ' ');
+    for (int i = counter; i <= 9; ++i) {
+        tmp += '0';
     }
 
-    pos = isoDate.find(' ');
-    std::string date = str.asString()->toNonGCUTF8StringData().substr(0, pos);
-    std::string time = str.asString()->toNonGCUTF8StringData().substr(pos + 1);
-    std::string year;
-    std::string month;
-    std::string week;
-    std::string day;
-    pos = date.find('-');
-
-    if (pos != std::string::npos) {
-        if (pos == 0 && date.length() > 4) {
-            if (date.substr(pos + 1).find('-') == 0) {
-                date.erase(std::remove(date.begin(), date.end(), '-'), date.end());
-                if (date.length() == 4) {
-                    month = date.substr(1, 2);
-                    day = date.substr(3, 2);
-                    year = "";
+    return { { "milliSecond", tmp.substr(0, 3) },
+             { "microSecond", tmp.substr(3, 3) },
+             { "nanoSecond", tmp.substr(6, 3) } };
+}
+
+void TemporalObject::offset(std::string& isoString)
+{
+    isoString = isoString.substr(1);
+    if (!(((isoString[0] == '0' || isoString[0] == '1') && isoString[1] >= '0' && isoString[1] <= '9') || (isoString[0] == '2' && isoString[1] >= 0 && isoString[1] <= '3'))) {
```

Comment:
The string is zero terminated, isn't it? So no need any length checks here.

Reply:
Yes it is.